src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs

```diff
@@ -65,30 +65,50 @@ internal static bool StructDependsOn(NamedTypeSymbol depends, NamedTypeSymbol on
             Debug.Assert((object)on != null);
             Debug.Assert(on.IsDefinition);
 
-            var hs = PooledHashSet<Symbol>.GetInstance();
-            StructDependsClosure(depends, hs, on);
+            var hs = PooledHashSet<NamedTypeSymbol>.GetInstance();
+            var typesWithCycle = PooledHashSet<NamedTypeSymbol>.GetInstance();
+            StructDependsClosure(depends, hs, typesWithCycle, ConsList<NamedTypeSymbol>.Empty.Prepend(on));
 
-            var result = hs.Contains(on);
+            var result = typesWithCycle.Contains(on);
+            typesWithCycle.Free();
             hs.Free();
 
             return result;
         }
 
-        private static void StructDependsClosure(NamedTypeSymbol type, HashSet<Symbol> partialClosure, NamedTypeSymbol on)
+        private static void StructDependsClosure(NamedTypeSymbol type, HashSet<NamedTypeSymbol> partialClosure, HashSet<NamedTypeSymbol> typesWithCycle, ConsList<NamedTypeSymbol> on)
         {
             Debug.Assert((object)type != null);
 
-            if ((object)type.OriginalDefinition == on)
+            if (typesWithCycle.Contains(type.OriginalDefinition))
+            {
+                return;
+            }
+
+            if (on.ContainsReference(type.OriginalDefinition))
             {
                 // found a possibly expanding cycle, for example
                 //     struct X<T> { public T t; }
                 //     struct W<T> { X<W<W<T>>> x; }
                 // while not explicitly forbidden by the spec, it should be.
-                partialClosure.Add(on);
+                typesWithCycle.Add(type.OriginalDefinition);
```

Comment:
>typesWithCycle.Add [](http://example.com/codeflow?start=16&length=18)

Are there any examples where `typesWithCycle` willl have more than one item?

Reply:
> Are there any examples where typesWithCycle willl have more than one item?

I see nothing preventing that. If that is what you are asking?

Reply:
If it's possible, it would be useful to have a test that has multiple entries in `typesWithCycle`, to prevent us from changing this to a single value in the future.